### ISSUE TITLE
Tests with DataProvider attribute not represented in ExecutionStarted->testSuite()->tests()

### ISSUE THREAD

**Issue by arderyp** [sebastianbergmann/phpunit#5212]
## ExecutionStartedSubscriber
```php
<?php declare(strict_types=1);

namespace App\Tests\Extension\Subscriber;

use Exception;
use PHPUnit\Event\TestRunner\ExecutionStarted;
use PHPUnit\Event\TestRunner\ExecutionStartedSubscriber as ExecutionStartedSubscriberInterface;

final class ExecutionStartedSubscriber implements ExecutionStartedSubscriberInterface
{
    public function notify(ExecutionStarted $event): void
    {
        $testSuite = $event->testSuite();
        echo "\tFound {$testSuite->count()} tests on TestSuite\n";
        $tests = $testSuite->tests();
        if ($tests->count()) {
            foreach ($tests as $test) {
                $class = $test->className();
                $method = $test->methodName();
                echo "\t- $class::$method\n";
            }
        } else {
            echo "\t- NO TESTS DETECTED\n\n";
        }
        die("TODO:REMOVE: exit early");
    }
}
```

## WithDataProviderTest
```php
<?php declare(strict_types=1);

namespace App\Tests\Javascript;

use PHPUnit\Framework\Attributes\DataProvider;

class WithDataProviderTest extends AbstractPantherTestCase
{
    public static function provider(): array
    {
        return [["foo"], ["bar"],];
    }

    #[DataProvider('provider')]
    public function testWith(string $string): void
    {
        $this->assertNotEmpty($string);
    }

}

```

## WithoutDataProviderTest
```php
<?php declare(strict_types=1);

namespace App\Tests\Javascript;

class WithoutDataProviderTest extends AbstractPantherTestCase
{
    public function testWithout(): void
    {
        $this->assertNotEmpty("foo");
    }

}
```

## WithAndWithoutDataProviderTest
```php
<?php declare(strict_types=1);

namespace App\Tests\Javascript;

use PHPUnit\Framework\Attributes\DataProvider;

class WithAndWithoutDataProviderTest extends AbstractPantherTestCase
{
    public static function provider(): array
    {
        return [["foo"], ["bar"],];
    }

    #[DataProvider('provider')]
    public function testWith(string $string): void
    {
        $this->assertNotEmpty($string);
    }

    public function testWithout(): void
    {
        $this->assertNotEmpty("foo");
    }
}

```

## Outputs
```bash
# PHPUnit 10.0.7 by Sebastian Bergmann and contributors.
#
# Runtime:       PHP 8.1.16
# Configuration: /home/arderyp/app/phpunit.xml

$./vendor/bin/phpunit tests/Javascript/WithDataProviderTest.php
	Found 2 tests on TestSuite
	- NO TESTS DETECTED

$ ./vendor/bin/phpunit tests/Javascript/WithoutDataProviderTest.php
	Found 1 tests on TestSuite
	- App\Tests\Javascript\WithoutDataProviderTest::testWithout


$ ./vendor/bin/phpunit tests/Javascript/WithAndWithoutDataProviderTest.php
	Found 3 tests on TestSuite
	- App\Tests\Javascript\WithAndWithoutDataProviderTest::testWithout

$ ./vendor/bin/phpunit
	Found 6 tests on TestSuite
	- NO TESTS DETECTED
```

## Output Expected
the output of `WithoutDataProviderTest` is correct, but the other outputs should be something along the lines of:
```bash
$./vendor/bin/phpunit tests/Javascript/WithDataProviderTest.php
	Found 2 tests on TestSuite
	- App\Tests\Javascript\WithDataProviderTest::testWith    ("foo")
 	- App\Tests\Javascript\WithDataProviderTest::testWith    ("bar")

$ ./vendor/bin/phpunit tests/Javascript/WithAndWithoutDataProviderTest.php
	Found 3 tests on TestSuite
	- App\Tests\Javascript\WithAndWithoutDataProviderTest::testWith    ("foo")
	- App\Tests\Javascript\WithAndWithoutDataProviderTest::testWith    ("foo")
	- App\Tests\Javascript\WithAndWithoutDataProviderTest::testWithout

$ ./vendor/bin/phpunit
	Found 6 tests on TestSuite
	- App\Tests\Javascript\WithAndWithoutDataProviderTest::testWith    ("foo")
	- App\Tests\Javascript\WithAndWithoutDataProviderTest::testWith    ("foo")
	- App\Tests\Javascript\WithAndWithoutDataProviderTest::testWithout
	- App\Tests\Javascript\WithDataProviderTest::testWith    ("foo")
 	- App\Tests\Javascript\WithDataProviderTest::testWith    ("bar")
	- App\Tests\Javascript\WithoutDataProviderTest::testWithout
```
The test count in the `Found N tests` is accurate, but the `tests()` property does not match.

**Comment by arderyp** [comment]
it seems there's a related issue here where, when using `--testsuite` or relying on `defaultTestSuite` in `phpunit.xml`, `ExecutionStarted->testSuite()->tests()` is always empty. This is demonstrated by my calls above to `phpunit` without any args.

**Comment by sebastianbergmann** [comment]
The [event log](https://github.com/sebastianbergmann/phpunit/blob/10.0/tests/end-to-end/event/test-data-provider.phpt) for a test that uses a data provider looks correct to me; that's a good start.

**Comment by sebastianbergmann** [comment]
This is what the `PHPUnit\Framework\TestSuite` object looks like that is passed to `PHPUnit\Event\TestSuite\TestSuite::fromTestSuite()` to build the `PHPUnit\Event\TestSuite\TestSuite` value object that is aggregated in the `PHPUnit\Event\TestRunner\ExecutionStarted`:

![testsuite](https://user-images.githubusercontent.com/25218/219935959-cd501f20-86a4-475c-afc5-3dfe346a2d8c.svg)

**TL;DR:** The `fromTestSuite()` method should have the information needed to create an `PHPUnit\Event\TestSuite\TestSuiteForTestMethodWithDataProvider` object with a correct array of tests.

**Comment by arderyp** [comment]
hey @sebastianbergmann thanks for taking the time to review!

I am not familiar with the event log, but just took a peak.  Are you saying that the expected output as documented by that file looks like what we want to see?

If so, I'm not sure where to go from here. I am happy to provide any further info or testing anything further if you point me in the right direction. Were you not able to duplicate my results with my code above?

**Comment by sebastianbergmann** [comment]
> Are you saying that the expected output as documented by that file looks like what we want to see?

The plain text event log is only useful to answer the question: are the right events emitted in the right order. It cannot be used to (at least not fully) answer the question of whether an event object has all information it is supposed to have.

In this case it is clear that the `TestSuite` object that is passed to the `ExecutionStarted` event is missing information. So yes, there is a bug. And yes, I will continute to look into this.

**Comment by arderyp** [comment]
Ahh, thanks for the clarification.  yes, the events seems to be firing properly in my rather extensive test suite and marginally complex Extension.

thanks for looking into it!  I started to dig on this one, but kind of hit a roadblock. I may have time to dig more, but I'm guessing you'll be able to spot it much quicker.

Anyways, thanks for the help.

P.S. PHPUnit 10 is awesome, and my testsuite runtime reduced by nearly 30%!

**Comment by sebastianbergmann** [comment]
After writing https://github.com/sebastianbergmann/phpunit/issues/5208#issuecomment-1435927344 I am not sure anymore whether there really is an issue here or if you are "just looking at the wrong event".

I will leave this ticket open and will come back to it with fresh eyes after the weekend.

**Comment by sebastianbergmann** [comment]
> PHPUnit 10 is awesome, and my testsuite runtime reduced by nearly 30%!

Thank you for the feedback. But as I already wrote in https://github.com/sebastianbergmann/phpunit/issues/5183#issuecomment-1423677465: I have no idea where such a performance improvement might be coming from.

**Comment by sebastianbergmann** [comment]
I do not think that there is anything to be done here.
